keps/sig-scheduling/4187-a-generic-scheduler-cache-extension/README.md

```diff
@@ -0,0 +1,777 @@
+![Class Diagram](class-diagram.png)
+
+#### New Interfaces
+
+- The ExtendedResource interface has been defined to manage the resource capacity/attributes, customize resource calculation model. The AddPod, RemovePod methods are invoked when a Pod is created on a node or removed from a node.  The AdmitPod method is invoked when a plugin checks whether the resource fits for a pod during the filtering stage. Scheduler plugin developers should implement this interface (e.g. XXXResource) to customize their extended resource availability calculation and check of resource fit for a new pod .
+
+- The ExtendedCache interface has been defined to set and get the ExtendedResource in the cache extension by node name.  
+
+- The ExtendedResource.Handle is in charge of managing resource changes and updating changes to ExtendedResource.  The Run method is called soon after the scheduler plugin is initialized. The GetExtendedResource method returns the corresponding ExtendedResource by node name. The scheduler plugin developers should implement this interface (e.g., XXXResourceHandle) to customize the update methods of resources’ available capacity.
+
+#### Modified Interfaces
+
+- framework Handle has been extended to return the ExtendedResource.Handle Registry for scheduler plugins to register handles.
+
+- The Cache interface has been extended to include the ability to retrieve extended cache by resource type. After the resource handle is registered, the extended cache of that resource is added to CacheImpl.
+
+#### New Data Structures
+
+- extendedCacheImpl has two fields, the "cacheImpl" field holds the implementation function for the cache interface, and the "key" field holds the resource type.  extendedCacheImpl implements the ExtendedCache.
+
+- HandleRegistry is a collection of ExtendedResource.Handles which consists of a Registry and a Register method. The key of the Registry is the resource type and the value of the Registry is the corresponding ExtendedResource.Handle to the resource type. The Register method is invoked when a new ExtendedResource.Handle is registered in the Handle registry.
+
+#### Modified Data Structures
+
+- A collection of ExtendedResources, where the key is the resource type, has been added as a new ExtendedResource field to the NodeInfo struct.
+
+- frameworkImpl has been extended to return the ExtendedResource.Handle Registry for the scheduler plugins to register ExtendedResource.Handle.
+
+- CacheImpl has been extended to manage a collection of ExtendedCaches which keeps track of the all the registered extendedCacheImpl instances (extendedCacheImpl implements the ExtendedCache interface which encapsulates the cacheImpl and a resource type).
+
+- frameworkOptions has been extended to include the HandleRegistry which would be passed to frameworkImpl to register the ExtendedResource.Handle.
```

Comment:
pls export all those fields from struct that expects to be accessed by the schedule plugin, e.g. `extendedResourceHandleRegistry` should be exported from `frameworkOptions` (see the struct definition from the diagram).

Reply:
resolved